### ISSUE TITLE
Warning in include/deal.II/matrix_free/mapping_info.templates.h

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#10902]
With gcc 10.2, I get this warning several dozen times:
```
/home/fac/g/bangerth/p/deal.II/1/dealii/include/deal.II/matrix_free/mapping_info.templates.h: In instantiation of ‘std::size_t dealii::internal::MatrixFreeFunctions::MappingInfoStorage<structdim, spacedim, Number, VectorizedArrayType>::QuadratureDescriptor::memory_consumption() const [with int structdim = 0; int spacedim = 1; Number = double; VectorizedArrayType = dealii::VectorizedArray<double, 2>; std::size_t = long unsigned int]’:
/home/fac/g/bangerth/p/deal.II/1/dealii/include/deal.II/base/memory_consumption.h:419:32:   required from ‘typename std::enable_if<(!(std::is_fundamental<_Tp>::value || std::is_pointer<_Tp>::value)), long unsigned int>::type dealii::MemoryConsumption::memory_consumption(const T&) [with T = dealii::internal::MatrixFreeFunctions::MappingInfoStorage<0, 1, double, dealii::VectorizedArray<double, 2> >::QuadratureDescriptor; typename std::enable_if<(!(std::is_fundamental<_Tp>::value || std::is_pointer<_Tp>::value)), long unsigned int>::type = long unsigned int]’
/home/fac/g/bangerth/p/deal.II/1/dealii/include/deal.II/base/memory_consumption.h:328:38:   required from ‘std::size_t dealii::MemoryConsumption::memory_consumption(const std::vector<Integer>&) [with T = dealii::internal::MatrixFreeFunctions::MappingInfoStorage<0, 1, double, dealii::VectorizedArray<double, 2> >::QuadratureDescriptor; std::size_t = long unsigned int]’
/home/fac/g/bangerth/p/deal.II/1/dealii/include/deal.II/matrix_free/mapping_info.templates.h:162:51:   required from ‘std::size_t dealii::internal::MatrixFreeFunctions::MappingInfoStorage<structdim, spacedim, Number, VectorizedArrayType>::memory_consumption() const [with int structdim = 0; int spacedim = 1; Number = double; VectorizedArrayType = dealii::VectorizedArray<double, 2>; std::size_t = long unsigned int]’
/home/fac/g/bangerth/p/deal.II/1/dealii/include/deal.II/base/memory_consumption.h:419:32:   required from ‘typename std::enable_if<(!(std::is_fundamental<_Tp>::value || std::is_pointer<_Tp>::value)), long unsigned int>::type dealii::MemoryConsumption::memory_consumption(const T&) [with T = dealii::internal::MatrixFreeFunctions::MappingInfoStorage<0, 1, double, dealii::VectorizedArray<double, 2> >; typename std::enable_if<(!(std::is_fundamental<_Tp>::value || std::is_pointer<_Tp>::value)), long unsigned int>::type = long unsigned int]’
/home/fac/g/bangerth/p/deal.II/1/dealii/include/deal.II/base/memory_consumption.h:328:38:   required from ‘std::size_t dealii::MemoryConsumption::memory_consumption(const std::vector<Integer>&) [with T = dealii::internal::MatrixFreeFunctions::MappingInfoStorage<0, 1, double, dealii::VectorizedArray<double, 2> >; std::size_t = long unsigned int]’
/home/fac/g/bangerth/p/deal.II/1/dealii/include/deal.II/matrix_free/mapping_info.templates.h:3133:54:   required from ‘std::size_t dealii::internal::MatrixFreeFunctions::MappingInfo<dim, Number, VectorizedArrayType>::memory_consumption() const [with int dim = 1; Number = double; VectorizedArrayType = dealii::VectorizedArray<double, 2>; std::size_t = long unsigned int]’
source/matrix_free/mapping_info.inst:78:48:   required from here
include/deal.II/matrix_free/mapping_info.templates.h:124:34: warning: comparison of unsigned expression in ‘< 0’ is always false [-Wtype-limits]
  124 |       for (unsigned int d = 0; d < structdim; ++d)
      |                                ~~^~~~~~~~~~~
```
We should probably try to address this somehow. Why are we instantiating for `structdim==0` at all?

**Comment by bangerth** [comment]
Similarly: 
```
/home/fac/g/bangerth/p/deal.II/1/dealii/source/base/quadrature.cc:92:30: warning: comparison of unsigned expression in ‘< 0’ is always false [-Wtype-limits]
   92 |   for (unsigned int i = 0; i < dim; ++i)
```
There are several more of this kind. Should we just disable `type-limis`?

**Comment by drwells** [comment]
This is the same warning I dealt with in #10854 and #10855. We usually end up instantiating these things when we instantiate face objects in 1D.

One possible fix would be to start using `if constexpr` (with a C++17 check, of course).

**Comment by peterrum** [comment]
> We should probably try to address this somehow. Why are we instantiating for structdim==0 at all?

It's the face in the context of 1D.

**Comment by kronbichler** [comment]
This particular warning was addressed by #11016, so let us close this for now.
